compiler-core/src/analyse.rs

```diff
@@ -1343,3 +1343,72 @@ fn get_compatible_record_fields<A>(
 
     compatible
 }
+
+/// Given a type, return a list of all the types it depends on
+fn get_type_dependencies(type_name: SmolStr, typ: &TypeAst) -> Vec<SmolStr> {
+    let mut deps = Vec::with_capacity(1);
+
+    match typ {
+        TypeAst::Var { name, .. } => deps.push(name.clone()),
+        TypeAst::Hole { name, .. } => deps.push(name.clone()),
+        TypeAst::Constructor {
+            name, arguments, ..
+        } => {
+            deps.push(name.clone());
+            for arg in arguments {
+                deps.extend(get_type_dependencies(type_name.clone(), arg))
+            }
+        }
+        TypeAst::Fn {
+            arguments, return_, ..
+        } => {
+            for arg in arguments {
+                deps.extend(get_type_dependencies(type_name.clone(), arg))
+            }
+            deps.extend(get_type_dependencies(type_name, return_))
+        }
+        TypeAst::Tuple { elems, .. } => {
+            for elem in elems {
+                deps.extend(get_type_dependencies(type_name.clone(), elem))
+            }
+        }
+    }
+
+    deps
+}
+
+/// Given a list of type aliases, return a list of the same type aliases sorted
+/// such that any type alias that depends on another type alias appears after
+/// the type alias it depends on.
+/// NOTE: Right now it's only sorting aliases but could actually be used to sort all the types later
+fn sorted_type_aliases(aliases: &Vec<TypeAlias<()>>) -> Result<Vec<TypeAlias<()>>, Error> {
+    let mut deps: Vec<(SmolStr, Vec<SmolStr>)> = Vec::with_capacity(aliases.len());
+
+    for alias in aliases {
+        deps.push((
+            alias.alias.clone(),
+            get_type_dependencies(alias.alias.clone(), &alias.type_ast),
+        ))
+    }
+
+    let sorted_deps = dep_tree::toposort_deps(deps).map_err(|err| {
+        let dep_tree::Error::Cycle(cycle) = err;
+
+        let mut cycle = cycle.clone();
+        cycle.reverse();
+
+        let first = cycle.first().unwrap();
```

Comment:
We can avoid cloning and reversing the vector by taking a reference to the last element.